Provide support for WTF::String in WebExtension JavaScript APIs

ItsJamie9494 opened the pull request:
#### 01f5315190c9623c7a03e449865a45d21554520e
<pre>
Provide support for WTF::String in WebExtension JavaScript APIs
<a href="https://bugs.webkit.org/show_bug.cgi?id=300606">https://bugs.webkit.org/show_bug.cgi?id=300606</a>

Reviewed by Timothy Hatcher.

This implements support for WTF::String in WebExtension’s JavaScript
APIs. While nothing is ported to C++, this provides a good starting point
for porting the APIs.

NSString is still supported in the JavaScript APIs, and most APIs still
use NSString at the moment, with the exception of the Action APIs, which
have been ported to using WTF::String.

No new tests are required, all existing WebExtension JavaScript tests
will test these changes.

* Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
(WebKit::toJSError):
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::toJSError): Deleted.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIActionCocoa.mm:
(WebKit::WebExtensionAPIAction::parseActionDetails):
(WebKit::WebExtensionAPIAction::getTitle):
(WebKit::WebExtensionAPIAction::setTitle):
(WebKit::WebExtensionAPIAction::getBadgeText):
(WebKit::WebExtensionAPIAction::setBadgeText):
(WebKit::WebExtensionAPIAction::getBadgeBackgroundColor):
(WebKit::WebExtensionAPIAction::setBadgeBackgroundColor):
(WebKit::WebExtensionAPIAction::isEnabled):
(WebKit::WebExtensionAPIAction::setIcon):
(WebKit::WebExtensionAPIAction::getPopup):
(WebKit::WebExtensionAPIAction::setPopup):
(WebKit::WebExtensionAPIAction::openPopup):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPortCocoa.mm:
(WebKit::WebExtensionAPIPort::postMessage):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntimeBase::reportError):
(WebKit::WebExtensionAPIRuntime::parseConnectOptions):
(WebKit::WebExtensionAPIRuntime::getURL):
(WebKit::WebExtensionAPIRuntime::getVersion):
(WebKit::WebExtensionAPIRuntime::runtimeIdentifier):
(WebKit::WebExtensionAPIRuntime::sendMessage):
(WebKit::WebExtensionAPIRuntime::connect):
(WebKit::WebExtensionAPIRuntime::sendNativeMessage):
(WebKit::WebExtensionAPIRuntime::connectNative):
(WebKit::WebExtensionAPIWebPageRuntime::sendMessage):
(WebKit::WebExtensionAPIWebPageRuntime::connect):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::sendMessage):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIAction.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIPort.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h:
* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
(WebKit::WebExtensionCallbackHandler::reportError):
(WebKit::toString):
(WebKit::toNSDictionary):
(WebKit::toJSValueRef):
(WebKit::deserializeJSONString):
(WebKit::serializeJSObject):
(WebKit::toJSError):
(WebKit::toJSString):
(WebKit::toNSString): Deleted.
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_generateImplementationFile):
(_installArgumentTypeExceptions):
(_installAutomaticExceptions):
(_platformType):
(_platformTypeConstructor):
(_platformTypeVariableDeclaration):
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/IDLAttributes.json:
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
(WebKit::WebExtensionControllerProxy::globalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::serviceWorkerGlobalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::addBindingsToWebPageFrameIfNecessary):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIAction.idl:

Canonical link: <a href="https://commits.webkit.org/301649@main">https://commits.webkit.org/301649@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae42568e11835e1ce7e2a5e7c34e68a44b4fbc38

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126540 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46185 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37108 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133429 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78214 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46819 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54723 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96295 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/64389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129488 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37451 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113168 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76837 "Found 1 new API test failure: WPE/TestWebsiteData:/webkit/WebKitWebsiteData/configuration (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36343 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76737 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107260 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31630 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135977 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53231 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40934 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104805 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53717 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109493 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104506 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49981 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28316 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/50645 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19806 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53151 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58964 "Found 11 new failures in WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm, WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52433 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55767 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54168 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -303,7 +310,7 @@ id parseJSON(API::Data& json, JSONOptionSet options, NSError **error)
             return nil;
 
-        return [value._toJSONString dataUsingEncoding:NSUTF8StringEncoding];
+        return [value._toJSONString.createNSString().autorelease() dataUsingEncoding:NSUTF8StringEncoding];
```
```suggestion
        return [value._toJSONString.createNSString() dataUsingEncoding:NSUTF8StringEncoding];
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
```diff
@@ -25,6 +25,8 @@
 
 #include "config.h"
 #include "WebExtensionUtilities.h"
+
```
```suggestion
```

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -283,7 +283,14 @@ id parseJSON(API::Data& json, JSONOptionSet options, NSError **error)
+        if (jsonString.isNull())
+            return nil;
+
+        return jsonString.createNSString().autorelease();
```
```suggestion
        return nsStringNilIfNull(jsonString);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
```diff
@@ -426,27 +444,27 @@ JSValueRef toJSValueRef(JSContextRef context, id object)
+        return nullString();
 
-    return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+    return String([[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding]);
```
Construction is implicit.

```suggestion
    return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
```diff
@@ -426,27 +444,27 @@ JSValueRef toJSValueRef(JSContextRef context, id object)
 {
     // This double-JSON approach works best since it avoids JSC's Cocoa object conversion, which can produce JSValue's that NSJSONSerialization can't convert.
-    auto* data = [self._toJSONString dataUsingEncoding:NSUTF8StringEncoding];
+    auto* data = [self._toJSONString.createNSString().autorelease() dataUsingEncoding:NSUTF8StringEncoding];
```
```suggestion
    auto* data = [self._toJSONString.createNSString() dataUsingEncoding:NSUTF8StringEncoding];
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
```diff
@@ -749,26 +755,39 @@ EOF
+EOF
+            } elsif ($needsExceptionString && $isVoidReturn) {
+                push(@contents, <<EOF);
+    String exceptionString = nullString();
```
```suggestion
    String exceptionString;
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
```diff
@@ -35,13 +35,11 @@
-- (NSString *)_toJSONString;
-- (NSString *)_toSortedJSONString;
+- (String)_toJSONString;
+- (String)_toSortedJSONString;
```
I don't think this makes sense, since this is an ObjC class. Why not keep the results of these as `NSString`? You will need to move to another type besides `JSValue` to use these in pure C++ anyway.

ItsJamie9494 replied:
I was worried about mixing these here. I described this is a different comment, but I think what we can do is we can use JSValueRef for most APIs, and have some extra functions to cover what JSValueRef can't do + what JSValue here does (such as checking function ,dictionary, regexp). For this PR, though, I'll just drop the JSValue changes

xeenon commented on Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
```diff
@@ -65,7 +65,11 @@
     auto globalObject = JSContextGetGlobalObject(context);
 
-    auto namespaceObject = JSObjectGetProperty(context, globalObject, toJSString("browser").get(), nullptr);
+    JSRetainPtr<JSStringRef> browserString = toJSString("browser");
```
```suggestion
    JSRetainPtr browserString = toJSString("browser");
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
```diff
@@ -85,8 +89,10 @@
-    JSObjectSetProperty(context, globalObject, toJSString("chrome").get(), namespaceObject, kJSPropertyAttributeNone, nullptr);
+    JSObjectSetProperty(context, globalObject, browserString.get(), namespaceObject, kJSPropertyAttributeNone, nullptr);
+
+    if (JSRetainPtr<JSStringRef> chromeString = toJSString("chrome"))
```
```suggestion
    if (JSRetainPtr chromeString = toJSString("chrome"))
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
```diff
@@ -100,30 +106,39 @@
     auto globalObject = JSContextGetGlobalObject(context);
 
-    auto namespaceObject = JSObjectGetProperty(context, globalObject, toJSString("browser").get(), nullptr);
+    JSRetainPtr<JSStringRef> browserString = toJSString("browser");
```
```suggestion
    JSRetainPtr browserString = toJSString("browser");
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
```diff
@@ -100,30 +106,39 @@
-    JSObjectSetProperty(context, globalObject, toJSString("browser").get(), namespaceObject, kJSPropertyAttributeNone, nullptr);
-    JSObjectSetProperty(context, globalObject, toJSString("chrome").get(), namespaceObject, kJSPropertyAttributeNone, nullptr);
+    JSObjectSetProperty(context, globalObject, browserString.get(), namespaceObject, kJSPropertyAttributeNone, nullptr);
+    if (JSRetainPtr<JSStringRef> chromeString = toJSString("chrome"))
```
```suggestion
    if (JSRetainPtr chromeString = toJSString("chrome"))
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
```diff
@@ -100,30 +106,39 @@
     auto globalObject = JSContextGetGlobalObject(context);
 
-    auto namespaceObject = JSObjectGetProperty(context, globalObject, toJSString("browser").get(), nullptr);
+    JSRetainPtr<JSStringRef> browserString = toJSString("browser");
```
```suggestion
    JSRetainPtr browserString = toJSString("browser");
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
```diff
@@ -275,18 +278,20 @@ id toNSObject(JSContextRef context, JSValueRef valueRef, Class containingObjects
 
     case NullOrEmptyString::NullStringAsEmptyString:
-        return JSValueMakeString(context, toJSString(string).get());
+        if (JSRetainPtr<JSStringRef> stringRef = toJSString(string))
```
```suggestion
        if (JSRetainPtr stringRef = toJSString(string))
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
```diff
@@ -314,49 +319,62 @@ JSValueRef toJSValueRef(JSContextRef context, NSURL *url, NullOrEmptyString null
+        return nullString();
+    }
+
+    return String::fromUTF8(stringBuffer);
```
```suggestion
    Vector<char> buffer(JSStringGetMaximumUTF8CStringSize(string));
    JSStringGetUTF8CString(string, buffer.mutableSpan().data(), buffer.size());
    return String::fromUTF8(buffer.span().data());
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
```diff
@@ -314,49 +319,62 @@ JSValueRef toJSValueRef(JSContextRef context, NSURL *url, NullOrEmptyString null
 
-    if (JSValueRef value = JSValueMakeFromJSONString(context, toJSString(jsonString).get()))
-        return value;
+    if (JSRetainPtr<JSStringRef> string = toJSString(jsonString)) {
```
```suggestion
    if (JSRetainPtr string = toJSString(jsonString)) {
```

xeenon reviewed (changes requested):
Looking good. By biggest request is around the `JSValue` changes for `_toJSONString` and friends. I think this can be reverted from the this PR to be better addressed with a plan to get off `JSValue`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPortCocoa.mm:
```diff
@@ -140,7 +140,7 @@
     RELEASE_LOG_DEBUG(Extensions, "Sent port message for channel %{public}llu from %{public}@ world", channelIdentifier().toUInt64(), toDebugString(contentWorldType()).createNSString().get());
 
-    WebProcess::singleton().send(Messages::WebExtensionContext::PortPostMessage(contentWorldType(), targetContentWorldType(), owningPageProxyIdentifier(), channelIdentifier(), message), extensionContext().identifier());
+    WebProcess::singleton().send(Messages::WebExtensionContext::PortPostMessage(contentWorldType(), targetContentWorldType(), owningPageProxyIdentifier(), channelIdentifier(), message.createNSString().get()), extensionContext().identifier());
```
Not needed, IPC expects `String` not `NSString`. You are double converting here.

ItsJamie9494 replied:
Ah, ok, didn't realise it was implicitly converting here

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -182,16 +183,16 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
+String WebExtensionAPIRuntime::getVersion()
 {
-    return objectForKey<NSString>(extensionContext().manifest(), versionKey);
+    return String(objectForKey<NSString>(extensionContext().manifest(), versionKey));
```
Converts for free.
```suggestion
    return objectForKey<NSString>(extensionContext().manifest(), versionKey);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -332,7 +333,7 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
     };
 
-    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::RuntimeSendMessage(extensionID, messageJSON, senderParameters), [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<String, WebExtensionError>&& result) {
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::RuntimeSendMessage(extensionID.createNSString().autorelease(), messageJSON.createNSString().autorelease(), senderParameters), [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<String, WebExtensionError>&& result) {
```
Not needed for IPC.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -370,7 +371,7 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
     Ref port = WebExtensionAPIPort::create(*this, webPageProxyIdentifier, WebExtensionContentWorldType::Main, resolvedName);
 
-    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::RuntimeConnect(extensionID, port->channelIdentifier(), resolvedName, senderParameters), [=, this, protectedThis = Ref { *this }, globalContext = JSRetainPtr { JSContextGetGlobalContext(context) }](Expected<void, WebExtensionError>&& result) {
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::RuntimeConnect(extensionID.createNSString().autorelease(), port->channelIdentifier(), resolvedName, senderParameters), [=, this, protectedThis = Ref { *this }, globalContext = JSRetainPtr { JSContextGetGlobalContext(context) }](Expected<void, WebExtensionError>&& result) {
```
Ditto, IPC.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -381,11 +382,11 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
     // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/runtime/sendNativeMessage
 
-    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::RuntimeSendNativeMessage(applicationID, messageJSON), [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<String, WebExtensionError>&& result) {
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::RuntimeSendNativeMessage(applicationID.createNSString().autorelease(), messageJSON.createNSString().autorelease()), [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<String, WebExtensionError>&& result) {
```
Ditto, IPC.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -395,13 +396,13 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
     // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/runtime/connectNative
 
-    Ref port = WebExtensionAPIPort::create(*this, webPageProxyIdentifier, WebExtensionContentWorldType::Native, applicationID);
+    Ref port = WebExtensionAPIPort::create(*this, webPageProxyIdentifier, WebExtensionContentWorldType::Native, applicationID.createNSString().autorelease());
```
Not needed, `WebExtensionAPIPort` takes a `String`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -395,13 +396,13 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
+    Ref port = WebExtensionAPIPort::create(*this, webPageProxyIdentifier, WebExtensionContentWorldType::Native, applicationID.createNSString().autorelease());
 
-    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::RuntimeConnectNative(applicationID, port->channelIdentifier(), webPageProxyIdentifier), [=, this, protectedThis = Ref { *this }, globalContext = JSRetainPtr { JSContextGetGlobalContext(context) }](Expected<void, WebExtensionError>&& result) {
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::RuntimeConnectNative(applicationID.createNSString().autorelease(), port->channelIdentifier(), webPageProxyIdentifier), [=, this, protectedThis = Ref { *this }, globalContext = JSRetainPtr { JSContextGetGlobalContext(context) }](Expected<void, WebExtensionError>&& result) {
```
Ditto, IPC.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -437,7 +438,7 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
     };
 
-    RefPtr destinationExtensionContext = page.webExtensionControllerProxy()->extensionContext(extensionID);
+    RefPtr destinationExtensionContext = page.webExtensionControllerProxy()->extensionContext(extensionID.createNSString().autorelease());
```
Not needed, already expects `String`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -447,7 +448,7 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
     }
 
-    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::RuntimeWebPageSendMessage(extensionID, messageJSON, senderParameters), [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<String, WebExtensionError>&& result) {
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::RuntimeWebPageSendMessage(extensionID.createNSString().autorelease(), messageJSON.createNSString().autorelease(), senderParameters), [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<String, WebExtensionError>&& result) {
```
Ditto, IPC.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -483,7 +484,7 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
     };
 
-    RefPtr destinationExtensionContext = page.webExtensionControllerProxy()->extensionContext(extensionID);
+    RefPtr destinationExtensionContext = page.webExtensionControllerProxy()->extensionContext(extensionID.createNSString().autorelease());
```
Ditto, not needed.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -497,7 +498,7 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
     Ref port = WebExtensionAPIPort::create(contentWorldType(), runtime(), *destinationExtensionContext, page.webPageProxyIdentifier(), WebExtensionContentWorldType::Main, resolvedName);
 
-    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::RuntimeWebPageConnect(extensionID, port->channelIdentifier(), resolvedName, senderParameters), [=, this, protectedThis = Ref { *this }, globalContext = JSRetainPtr { JSContextGetGlobalContext(context) }](Expected<void, WebExtensionError>&& result) {
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::RuntimeWebPageConnect(extensionID.createNSString().autorelease(), port->channelIdentifier(), resolvedName, senderParameters), [=, this, protectedThis = Ref { *this }, globalContext = JSRetainPtr { JSContextGetGlobalContext(context) }](Expected<void, WebExtensionError>&& result) {
```
Ditto, not needed.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
```diff
@@ -1006,7 +1006,7 @@ bool isValid(std::optional<WebExtensionTabIdentifier> identifier, NSString **out
     };
 
-    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::TabsSendMessage(tabIdentifer.value(), messageJSON, targetParameters, senderParameters), [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<String, WebExtensionError>&& result) {
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::TabsSendMessage(tabIdentifer.value(), messageJSON.createNSString().get(), targetParameters, senderParameters), [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<String, WebExtensionError>&& result) {
```
Ditto, not needed.

xeenon commented on Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
```diff
@@ -247,14 +248,16 @@ id toNSObject(JSContextRef context, JSValueRef valueRef, Class containingObjects
             continue;
 
-        auto *key = toNSString(propertyName.get());
+        auto key = toString(propertyName.get()).createNSString().get();
```
`*` since it is still ObjC.

```suggestion
        auto *key = toString(propertyName.get()).createNSString().get();
```

xeenon reviewed (changes requested):
Looking good, just some fixes for the double string conversions.

xeenon reviewed (approved):
Great!

ItsJamie9494 committed:
Provide support for WTF::String in WebExtension JavaScript APIs
https://bugs.webkit.org/show_bug.cgi?id=300606

Reviewed by Timothy Hatcher.

This implements support for WTF::String in WebExtension’s JavaScript
APIs. While nothing is ported to C++, this provides a good starting point
for porting the APIs.

NSString is still supported in the JavaScript APIs, and most APIs still
use NSString at the moment, with the exception of the Action APIs, which
have been ported to using WTF::String.

No new tests are required, all existing WebExtension JavaScript tests
will test these changes.

* Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
(WebKit::toJSError):
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::toJSError): Deleted.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIActionCocoa.mm:
(WebKit::WebExtensionAPIAction::parseActionDetails):
(WebKit::WebExtensionAPIAction::getTitle):
(WebKit::WebExtensionAPIAction::setTitle):
(WebKit::WebExtensionAPIAction::getBadgeText):
(WebKit::WebExtensionAPIAction::setBadgeText):
(WebKit::WebExtensionAPIAction::getBadgeBackgroundColor):
(WebKit::WebExtensionAPIAction::setBadgeBackgroundColor):
(WebKit::WebExtensionAPIAction::isEnabled):
(WebKit::WebExtensionAPIAction::setIcon):
(WebKit::WebExtensionAPIAction::getPopup):
(WebKit::WebExtensionAPIAction::setPopup):
(WebKit::WebExtensionAPIAction::openPopup):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPortCocoa.mm:
(WebKit::WebExtensionAPIPort::postMessage):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntimeBase::reportError):
(WebKit::WebExtensionAPIRuntime::parseConnectOptions):
(WebKit::WebExtensionAPIRuntime::getURL):
(WebKit::WebExtensionAPIRuntime::getVersion):
(WebKit::WebExtensionAPIRuntime::runtimeIdentifier):
(WebKit::WebExtensionAPIRuntime::sendMessage):
(WebKit::WebExtensionAPIRuntime::connect):
(WebKit::WebExtensionAPIRuntime::sendNativeMessage):
(WebKit::WebExtensionAPIRuntime::connectNative):
(WebKit::WebExtensionAPIWebPageRuntime::sendMessage):
(WebKit::WebExtensionAPIWebPageRuntime::connect):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::sendMessage):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIAction.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIPort.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h:
* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
(WebKit::WebExtensionCallbackHandler::reportError):
(WebKit::toString):
(WebKit::toNSDictionary):
(WebKit::toJSValueRef):
(WebKit::deserializeJSONString):
(WebKit::serializeJSObject):
(WebKit::toJSError):
(WebKit::toJSString):
(WebKit::toNSString): Deleted.
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_generateImplementationFile):
(_installArgumentTypeExceptions):
(_installAutomaticExceptions):
(_platformType):
(_platformTypeConstructor):
(_platformTypeVariableDeclaration):
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/IDLAttributes.json:
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
(WebKit::WebExtensionControllerProxy::globalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::serviceWorkerGlobalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::addBindingsToWebPageFrameIfNecessary):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIAction.idl:

Canonical link: https://commits.webkit.org/301649@main
